recipes-ni/ni-base-system-image-tests/files/fs_permissions_known.py

```diff
@@ -0,0 +1,317 @@
+                            }
+                        }
+                    },
+                    'source': system_link('build')
+                }
+            }
+        }
+    }
+
+# Use bash to resolve strings like path globs and $() expressions.
+def bash(path):
+    cmd = [
+        'bash',
+        '-O', 'extglob',
+        '-O', 'globstar',
+        '-c', f'echo {path}' ]
+    return fs_permissions_shared.run_cmd(cmd).strip().split(' ')
+
+# Convert the tree structure into a list of paths and functions to check their permissions
+def known_permissions():
+    def permissions_list(tree, path, perms):
+        for file in tree:
+            if type(tree[file]) is dict:
+                perms = permissions_list(tree[file], f'{path}/{file}', perms)
+            elif file == '.':
+                perms[bash(path if path != '' else '/')[0]] = tree[file]
+            else:
+                subfiles = bash(f'{path}/{file}')
+                for subfile in subfiles:
+                    perms[subfile] = tree[file]
+        return perms
+    return permissions_list(known_permissions_tree(), '', {})
+
+#------- Functions for the tree to check file permissions ------
+
+FT_DIR = 0
+FT_REG = 1
+FT_LNK = 2
+
+# Report errors when a kind of thing is different than expected
+def log_mismatch(kind, exp, act, path, logger, md5sum):
+    md5sum.update(f'{kind}: {exp} != {act}'.encode('utf-8'))
+    logger.log(f'ERROR: Expected {kind} \'{exp}\' but got \'{act}\' for path {path}.')
+
+# Main check for file/directory permissions
+def permissions(mode, user, group, file_type):
+    def ret(path, stats, logger, md5sum):
+        nonlocal mode, user, group, file_type
+        if file_type == FT_DIR and not stat.S_ISDIR(stats.st_mode):
+            log_mismatch(
+                'file type', 'dir', 'reg' if file_type == FT_REG else 'lnk', path, logger, md5sum
+            )
+            return False
+        if file_type == FT_REG and not stat.S_ISREG(stats.st_mode):
+            log_mismatch(
+                'file type', 'reg', 'dir' if file_type == FT_DIR else 'lnk', path, logger, md5sum
+            )
+            return False
+        if file_type == FT_LNK and not stat.S_ISLNK(stats.st_mode):
+            log_mismatch(
+                'file type', 'lnk', 'dir' if file_type == FT_DIR else 'reg', path, logger, md5sum
+            )
+            return False
+        if mode != stat.S_IMODE(stats.st_mode):
+            log_mismatch('mode', oct(mode), oct(stat.S_IMODE(stats.st_mode)), path, logger, md5sum)
+            return False
+        if user != pwd.getpwuid(stats.st_uid).pw_name:
+            log_mismatch('user', user, pwd.getpwuid(stats.st_uid).pw_name, path, logger, md5sum)
+            return False
+        if group != grp.getgrgid(stats.st_gid).gr_name:
+            log_mismatch('group', group, grp.getgrgid(stats.st_gid).gr_name, path, logger, md5sum)
+            return False
+        return True
```

Comment:
Do we want to do early exit after the first mismatch, or provide all mismatches at once? Later on in `system_link` it appears to perform other checks even when the link target is a known mismatch.

```suggestion
        success = True
        if file_type == FT_DIR and not stat.S_ISDIR(stats.st_mode):
            log_mismatch(
                'file type', 'dir', 'reg' if file_type == FT_REG else 'lnk', path, logger, md5sum
            )
            success = False
        if file_type == FT_REG and not stat.S_ISREG(stats.st_mode):
            log_mismatch(
                'file type', 'reg', 'dir' if file_type == FT_DIR else 'lnk', path, logger, md5sum
            )
            success =False
        if file_type == FT_LNK and not stat.S_ISLNK(stats.st_mode):
            log_mismatch(
                'file type', 'lnk', 'dir' if file_type == FT_DIR else 'reg', path, logger, md5sum
            )
            success =False
        if mode != stat.S_IMODE(stats.st_mode):
            log_mismatch('mode', oct(mode), oct(stat.S_IMODE(stats.st_mode)), path, logger, md5sum)
            success =False
        if user != pwd.getpwuid(stats.st_uid).pw_name:
            log_mismatch('user', user, pwd.getpwuid(stats.st_uid).pw_name, path, logger, md5sum)
            success =False
        if group != grp.getgrgid(stats.st_gid).gr_name:
            log_mismatch('group', group, grp.getgrgid(stats.st_gid).gr_name, path, logger, md5sum)
            success =False
        return success
```